File: ruby/lib/ruby/utils/dates.rb
# A util class to operate dates
module Dates
    # Returns if a date exceeds a limit
    #
    # @param date [Time] date to compare with the current date
    # @param limit [Integer] difference limit in minutes between the dates
    #
    # Examples:
    #   Dates.is_greatter(Time.now - 120, 1)
    #     => true
    #
    #   Dates.is_greatter(Time.now + 120, 1)
    #     => false
    #
    # @return [Boolean] true if the difference doesn't exceeds the limit
    def self.is_greatter(date, limit)
        return self.get_diff_in_minutes(date) > limit
    end

    # Returns the difference between the current date and the given date
    #
    # @param date [Time] date to compare with the current date
    #
    # Examples:
    #   Dates.get_diff_in_minutes(Time.now - 120)
    #     => 2.00000006625
    #
    #   Dates.get_diff_in_minutes(Time.now + 120)
    #     => -1.9999999124
    #
    # @return [Float] The difference between the current date and the given date
    def self.get_diff_in_minutes(date)
        now = Time.now
        return self.diff_in_minutes(date, now)
    end

    private
    # Returns the difference between two dates
    #
    # @param inital_date [Time] Start date to compare
    # @param final_date [Time] Final date to compare
    #
    # Examples:
    #   The operation performed is final_date - inital_date
    #
    #   diff_in_minutes(Time.now, Time.now + 180)
    #     => 3.0000000196166665
    #
    #   diff_in_minutes(Time.now + 180, Time.now)
    #     => -2.9999999579166663
    #
    # @return [Float] The difference between the given dates
    def self.diff_in_minutes(init_date, final_date)
        return (final_date - init_date) / 60
    end
end

File: golang/cmd/src/utils/dates.go
// Package utils provide useful functions and utilities
// to process different types of data or perform different process
package utils

import "time"

// Dates is an useful struct that allow us
// to operate over dates and times.
type Dates struct{}

// IsGreatter returns if a date exceeds a limit.
// Obtain the date given minutes using the utils.Date{} GetDiffInMinutes method
// and copmare it with the limit given.
//
// The param date is the date to compare with the current date
// The param limit is the value that couldn't be exceeded
//
// Examples:
//		util.Dates{}.IsGreatter(time.Now().Add(time.Second * (-120)), 1)
//		  => true
//
//		util.Dates{}.IsGreatter(time.Now().Add(time.Second * +120), 1)
//        => false
//
// It returns true if the difference doesn't exceeds the limit, false otherwise
func (d Dates) IsGreatter(date time.Time, limit int32) bool {
	return d.GetDiffInMinutes(date) > limit
}

// GetDiffInMinutes returns the difference between two dates in terms of minutes
// It uses the utils.Date{} DiffInMinutes to operate in terms of minutes
//
// The param date is the date to compare with the current date
//
// Examples:
//		util.Dates{}.GetDiffInMinutes(time.Now().Add(time.Second * (-120)))
//		  => 2
//
//		util.Dates{}.GetDiffInMinutes(time.Now().Add(time.Second * 120))
//		  => -2
//
// It returns the difference between the current date and the given date
func (d Dates) GetDiffInMinutes(date time.Time) int32 {
	now := time.Now()
	return d.DiffInMinutes(date, now)
}

// DiffInMinutes returns the difference between two dates and parse the difference to minutes
//
// The param initial is the start date to compare
// The param final is the final date to compare
//
// Examples:
//		The operation performed is final_date - inital_date
//
//		util.Dates{}.DiffInMinutes(time.Now(), time.Now().Add(time.Second * 180)
//		  => 3
//
//		util.Dates{}.DiffInMinutes(time.Now().Add(time.Second * 180, time.Now())
//		  => -3
//
// It returns the difference between the given dates
func (d Dates) DiffInMinutes(initial time.Time, final time.Time) int32 {
	// Divide 1000000 to get milliseconds -> (A millisecond is 1_000_000 microseconds)
	// Divide 1000 to get seconds -> (a second is 1000 milliseconds)
	// Divide 60 to get minutes -> (a minute is 60 seconds)
	// Divide 60 to get hours -> (a hour is 60 minutes)
	// Divide 24 to get days -> (a day us 24 hours)
	return int32(int64(final.Sub(initial)) / 1000000 / 1000 / 60)
}
